source/agora/consensus/EnrollmentManager.d

```diff
@@ -532,6 +532,92 @@ public class EnrollmentManager
         return false;
     }
 
+    /***************************************************************************
+
+        Get all the valid validators
+
+        Params:
+            validators = will be filled with all the validators during
+                their validation cycles
+
+        Returns:
+            Return true if there is no error in getting validators
+
+    ***************************************************************************/
+
+    public bool getValidators (out Enrollment[] validators) @safe nothrow
+    {
+        try
+        {
+            () @trusted {
+                auto results = this.db.execute("SELECT val FROM validator_set" ~
+                    " WHERE enrolled_height is not null");
+                foreach (row; results)
+                {
+                    auto enroll =
+                        deserializeFull!Enrollment(row.peek!(ubyte[])(0));
+                    validators ~= enroll;
+                }
+            }();
+        }
+        catch (Exception ex)
+        {
+            log.error("Database operation error, exception:{}", ex);
+            return false;
+        }
+
+        return true;
+    }
+
+    /***************************************************************************
+
+        Set current block height. The enrollment manager can weed out expired
+            validators from the set based on the block height.
+
+        Params:
+            block_height = current block height
+
+    ***************************************************************************/
+
+    public void setCurrentBlockHeight (ulong block_height) @safe nothrow
+    {
+        Enrollment[string] validators;
+        ulong[string] enrolled_heights;
+        try
+        {
+            () @trusted {
+                auto results = this.db.execute("SELECT key, val, enrolled_height" ~
+                    " FROM validator_set WHERE enrolled_height is not null");
+                foreach (row; results)
+                {
+                    auto key = row[0].as!(string);
+                    validators[key] =
+                        deserializeFull!Enrollment(row.peek!(ubyte[])(1));
+                    enrolled_heights[key] = row[2].as!(ulong);
+                }
+            }();
+        }
+        catch (Exception ex)
+        {
+            log.error("Database operation error, exception:{}", ex);
+        }
+
+        try
+        {
+            // weed out expired validator from the set
+            foreach (ref key, height; enrolled_heights)
+            {
+                if (block_height >= height + ValidatorCycle)
```

Comment:
I don't quite understand why you would do this over `("DELETE FROM validator_set WHERE enrolled_height <= ?", block_height - ValidatorCycle)` ?

Reply:
Your suggestion is quite simple. I was not able to realize that the SQL statement can be made to the extent.